modules/library/main/src/main/java/org/geotools/feature/visitor/GroupByInternalVisitor.java

```diff
@@ -0,0 +1,166 @@
+/*
+ *    GeoTools - The Open Source Java GIS Toolkit
+ *    http://geotools.org
+ *
+ *    (C) 2016, Open Source Geospatial Foundation (OSGeo)
+ *
+ *    This library is free software; you can redistribute it and/or
+ *    modify it under the terms of the GNU Lesser General Public
+ *    License as published by the Free Software Foundation;
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.geotools.feature.visitor;
+
+import org.geotools.util.Converters;
+import org.geotools.util.UnsupportedImplementationException;
+import org.opengis.filter.expression.Expression;
+
+/**
+ * This Enum contains all the visitors that can be used with the group by visitor.
+ * Each visitor supported by the group by visitor has to provide a consistent way to
+ * create a new instance and if the visitor result can be optimized a way to wrap
+ * the raw result.
+ */
+public enum GroupByInternalVisitor {
```

Comment:
Thanks for breaking this out into seperate class, I will refactor this to:

````
public enum Aggregate {
   AVERAGE { ... },
   COUNT {...},
   ...
}
````

The enum is not acting as a visitor (since feature collection does not "accept" it).